### PR TITLE
[WIP] Trainer

### PR DESCRIPTION
**Pull request by DeNeutoy**

### CODE REVIEW

File: allennlp/training/regularizers.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+class RegularizerApplicator:
+    """
+    Recursively applies a list of regularizers to a Module
+    and it's children. A regularizer will only be applied to
+    a module if its regex matches the module class name.
+    """
+    def __init__(self, regluarizers: List[Regularizer]):
+
+        self._regularizers = regluarizers
```
Comment (DeNeutoy) [inline allennlp/training/regularizers.py]: Thanks

File: allennlp/training/trainer.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import torch
+
+from ..data.dataset import Dataset
+from ..common.params import Params
+from ..common.checks import ConfigurationError
+from .optimizers import get_optimizer_from_params
+from ..data.iterators import Iterator, concrete_iterators
+
+class Trainer:
```
Comment (DeNeutoy) [inline allennlp/training/trainer.py]: Yeah this isn't finished - it will have more features, no fear!

Eventually this will handle:
- Training
- Evaluation on held out data
- Saving and loading a model checkpoint, which will include optimizer state and training metrics
- Model parallism/moving data and models to GPUs.
- Callbacks which effect training, such as saving models, learning rate decay, logging metrics etc.

**Comment by DeNeutoy** [comment]
Splitting this into smaller pieces.
